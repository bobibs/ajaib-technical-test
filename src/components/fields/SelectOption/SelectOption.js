import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import ImageBox from '../../elements/ImageBox';
import styles from './styles.module.css';

export default function SelectOption(props) {
  const { className, onClick, optionList, resetValue } = props;
  const [open, setOpen] = useState(false);
  const [value, setValue] = useState('');

  const classes = [styles.root, className].filter(Boolean).join(' ');
  const handleClick = () => setOpen(!open);
  const handleClickOption = (val) => {
    setValue(val);
    setOpen(false);
    onClick(val.length ? val : '');
  };

  const list = [{ label: 'All', value: '' }, ...optionList];

  useEffect(() => {
    if (resetValue) {
      setValue('');
    }
  }, [resetValue]);

  return (
    <div className={classes}>
      <button className={styles.select} onClick={handleClick}>
        <div className={styles.value}>{value.length ? value : 'All'}</div>
        <ImageBox
          className={styles.icon}
          src={open ? 'caret-down-blue' : 'caret-up-blue'}
        />
      </button>
      {open ? (
        <div className={styles.dropdown}>
          {list.map((i, idx) => (
            <div
              className={styles.option}
              key={idx}
              onClick={() => handleClickOption(i.value)}
            >
              {i.label}
            </div>
          ))}
        </div>
      ) : (
        ''
      )}
    </div>
  );
}

SelectOption.defaultProps = {
  className: '',
  onClick: () => {},
  optionList: [],
  resetValue: false,
};

SelectOption.propTypes = {
  className: PropTypes.string,
  onClick: PropTypes.func,
  optionList: PropTypes.array,
  resetValue: PropTypes.bool,
};
