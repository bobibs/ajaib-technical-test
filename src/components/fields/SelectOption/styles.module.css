.select {
    height: 2rem;
    width: 10rem;
    margin: 0 0 0.5rem 0;
    padding: 0 0.5rem;
    border: 0.05rem solid #888888;
    border-radius: 0.25rem;
    display: flex;
    align-items: center;
    justify-content: space-between;
    cursor: pointer;
    background: #fff;
}

.value {
    text-transform: capitalize;
}

.icon {
    height: 1rem;
    width: 1rem;
}

.dropdown {
    position: absolute;
    background: #f4f4f4;
    border: 0.05rem solid #888888;
    border-radius: 0.25rem;
    width: 10rem;
    z-index: 1;
}

.dropdown > div:first-child {
    border-radius: 0.25rem 0.25rem 0 0;
}

.dropdown > div:last-child {
    border-radius: 0 0 0.25rem 0.25rem;
}

.option {
    cursor: pointer;
    background: #fff;
    padding: 0.5rem;
}

.option:hover {
    background: #d4d4d4;
}