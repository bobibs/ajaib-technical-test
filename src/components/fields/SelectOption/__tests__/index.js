import React, { useEffect, useState } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import SelectOption from '../SelectOption';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useEffect: jest.fn(),
}));

const open = false;
const setOpen = jest.fn();
const value = '';
const setValue = jest.fn();

describe('src/components/elements/SelectOption', () => {
  beforeEach(() => {
    useState.mockImplementation(() => [open, setOpen]);
    useState.mockImplementation(() => [value, setValue]);
    useEffect.mockImplementation((fn) => fn());
  });

  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<SelectOption />);
    SelectOption.defaultProps.onClick();
    expect(tree).toMatchSnapshot();
  });

  test('function handleClick', () => {
    const onClick = jest.fn();

    useState.mockImplementationOnce(() => [true, setOpen]);
    useState.mockImplementationOnce(() => ['male', setValue]);

    const result = SelectOption({ resetValue: true, optionList: [], onClick });
    result.props.children[0].props.onClick();
    expect(setOpen).toHaveBeenCalledWith(false);

    console.log(result.props.children[1].props.children[0].props.onClick);
    result.props.children[1].props.children[0].props.onClick('male');
    expect(onClick).toHaveBeenCalled();
  });
});
