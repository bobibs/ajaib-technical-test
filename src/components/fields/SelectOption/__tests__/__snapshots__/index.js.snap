// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/SelectOption render 1`] = `
<div
  className="root"
>
  <button
    className="select"
    onClick={[Function]}
  >
    <div
      className="value"
    >
      All
    </div>
    <ImageBox
      className="icon"
      format="svg"
      onClick={[Function]}
      src="caret-up-blue"
    />
  </button>
  
</div>
`;
